TLM/TLM/Util/Extensions/NetSegmentExtensions.cs

```diff
@@ -9,5 +53,130 @@ public static class NetSegmentExtensions {
+        /// <param name="laneTypeFilter">lane type filter, lanes must match this filter mask</param>
+        /// <param name="vehicleTypeFilter">vehicle type filter, lanes must match this filter mask</param>
+        /// <param name="reverse">if true, lanes are ordered from right to left (relative to the
+        ///     segment's start node / the given node), otherwise from left to right</param>
+        /// <param name="sort">if false, no sorting takes place
+        ///     regardless of <paramref name="reverse"/></param>
+        /// <returns>sorted list of lanes for the given segment</returns>
+        public static IList<LanePos> GetSortedLanes(
+            this ref NetSegment netSegment,
+            bool? startNode,
+            NetInfo.LaneType? laneTypeFilter = null,
+            VehicleInfo.VehicleType? vehicleTypeFilter = null,
+            bool reverse = false,
+            bool sort = true) {
+            // TODO refactor together with getSegmentNumVehicleLanes, especially the vehicle type and lane type checks
+            NetManager netManager = Singleton<NetManager>.instance;
+            var laneList = new List<LanePos>();
+
+            bool inverted = (netSegment.m_flags & NetSegment.Flags.Invert) != NetSegment.Flags.None;
+
+            NetInfo.Direction? filterDir = null;
+            NetInfo.Direction sortDir = NetInfo.Direction.Forward;
+
+            if (startNode != null) {
+                filterDir = (bool)startNode
+                                ? NetInfo.Direction.Backward
+                                : NetInfo.Direction.Forward;
+                filterDir = inverted
+                                ? NetInfo.InvertDirection((NetInfo.Direction)filterDir)
+                                : filterDir;
+                sortDir = NetInfo.InvertDirection((NetInfo.Direction)filterDir);
+            } else if (inverted) {
+                sortDir = NetInfo.Direction.Backward;
+            }
+
+            if (reverse) {
+                sortDir = NetInfo.InvertDirection(sortDir);
+            }
+
+            NetInfo segmentInfo = netSegment.Info;
+            uint curLaneId = netSegment.m_lanes;
+            byte laneIndex = 0;
+
+            while (laneIndex < segmentInfo.m_lanes.Length && curLaneId != 0u) {
+                NetInfo.Lane laneInfo = segmentInfo.m_lanes[laneIndex];
+                if ((laneTypeFilter == null ||
+                     (laneInfo.m_laneType & laneTypeFilter) != NetInfo.LaneType.None) &&
+                    (vehicleTypeFilter == null || (laneInfo.m_vehicleType & vehicleTypeFilter) !=
+                     VehicleInfo.VehicleType.None) &&
+                    (filterDir == null ||
+                     segmentInfo.m_lanes[laneIndex].m_finalDirection == filterDir)) {
+                    laneList.Add(
+                        new LanePos(
+                            curLaneId,
+                            laneIndex,
+                            segmentInfo.m_lanes[laneIndex].m_position,
```

Comment:
`laneInfo.m_position` ?

Reply:
First line in the `while` block:
```csharp
NetInfo.Lane laneInfo = segmentInfo.m_lanes[laneIndex];
```

So I still think valid to replace L126: `segmentInfo.m_lanes[laneIndex].m_finalDirection` with `laneInfo.m_finalDirection`

and L131: `segmentInfo.m_lanes[laneIndex].m_position` with `laneInfo.m_position`